test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,461 @@
+import os
+import tempfile
+from string import Template
+import copy
+import unittest
+import warnings
+import inspect
+
+import torch
+from torch._six import PY2
+import common_utils as common
+import common_nn
+from common_cuda import TEST_CUDA
+import torch.utils.cpp_extension
+from cpp_api_parity import sample_module, torch_nn_modules, TorchNNTestParams, set_has_parity
+
+
+torch_nn_has_parity = set([
+    'SampleModule',
+])
+
+TORCH_NN_MODULE_COMMON_TEST_HARNESS = """\n
+#include <torch/script.h>
+
+const char * const parity_test_error_msg = "Parity test failed";
+
+bool check_tensor_equality(const torch::Tensor& tensor1, const torch::Tensor& tensor2) {
+  return tensor1.sizes().vec() == tensor2.sizes().vec() && tensor1.allclose(tensor2);
+}
+
+bool check_ivalue_equality(const c10::IValue& ivalue1, const c10::IValue& ivalue2) {
+  if (ivalue1.tagKind() != ivalue2.tagKind()) {
+    AT_ERROR("Value type mismatch: ", "ivalue1: ", ivalue1.tagKind(), ", ivalue2: ", ivalue2.tagKind());
+  }
+  if (ivalue1.isInt()) {
+    return ivalue1.toInt() == ivalue2.toInt();
+  } else if (ivalue1.isDouble()) {
+    return ivalue1.toDouble() == ivalue2.toDouble();
+  } else if (ivalue1.isBool()) {
+    return ivalue1.toBool() == ivalue2.toBool();
+  } else if (ivalue1.isString()) {
+    return ivalue1.toString() == ivalue2.toString();
+  } else if (ivalue1.isTensor()) {
+    return check_tensor_equality(ivalue1.toTensor(), ivalue2.toTensor());
+  } else {
+    AT_ERROR("Unsupported value type: ", ivalue1.tagKind());
+  }
+}
+"""
+
+CHECK_MODULE_PARAM_EQUALITY = Template("""\
+TORCH_CHECK(
+  check_tensor_equality(${script_module_prefix}.get_parameter("${param_name}"), ${cpp_module_prefix}->${param_name}),
```

Comment:
For `${cpp_module_prefix}->${param_name}`, I actually want to test that `param_name` exists as an attribute of the C++ module, not just as a registered parameter. Because there can be cases where `cpp_module->named_parameters()["weight"]` exists (e.g. `register_parameter("weight", ...)` is called) but `cpp_module->weight` doesn't exist, and I would like to prevent that from happening.

And we probably can't use the helper function and pass in `param_name` because we need to generate the `${cpp_module_prefix}->${param_name}` statement directly. :(

Reply:
Oh, I see, that makes sense. Thanks for the clarification